site/docs/v1/tech/tutorials/integrate-laravel-api.md

```diff
@@ -0,0 +1,245 @@
+
+```diff
+{% remote_include https://github.com/FusionAuth/fusionauth-example-laravel-api/commit/e4298a10437ff396fe3dfc02c3bcf85f310921a0.diff %}
+```
+
+Change the created migration at `database/migrations/2014_10_12_000000_create_users_table.php`.
+
+```diff
+{% remote_include https://github.com/FusionAuth/fusionauth-example-laravel-api/commit/9c670115fa991d2219120a88ccaaf94b9e5feab1.diff %}
+```
+
+You may remove all other migrations, as you won't need them.
+
+```shell
+rm database/migrations/2014_10_12_100000_create_password_reset_tokens_table.php
+rm database/migrations/2019_08_19_000000_create_failed_jobs_table.php
+rm database/migrations/2019_12_14_000001_create_personal_access_tokens_table.php
+```
+
+Run the remaining migrations to create the necessary tables in your database.
+
+```shell
+./vendor/bin/sail artisan migrate
+```
+
+### Adding Files
+
+By default, Laravel only allows JWTs that correspond to existing users in your database, but one of the greatest benefits of using FusionAuth is to have a single source of truth of user management. So, you would want your API to automatically provision new users when it receives a trusted JWT from FusionAuth. Even though [`tymon/jwt-auth`](https://github.com/tymondesigns/jwt-auth) is a great library, it doesn't provide an easy way of doing these kinds of customizations, so you'll have to create some classes to do them.
+
+First, extend the default [User Provider](https://laravel.com/docs/10.x/authentication#adding-custom-user-providers) to create new users when receiving valid JWTs from FusionAuth.
+
+```php
+{% remote_include https://raw.githubusercontent.com/FusionAuth/fusionauth-example-laravel-api/main/laravel/app/FusionAuth/Providers/FusionAuthEloquentUserProvider.php %}
+```
+
+Now, add an [Authentication Guard](https://laravel.com/docs/10.x/authentication#adding-custom-guards) in `app/FusionAuth/FusionAuthJWTGuard.php` to call that custom method from the User Provider created above.
+
+```php
+{% remote_include https://raw.githubusercontent.com/FusionAuth/fusionauth-example-laravel-api/main/laravel/app/FusionAuth/FusionAuthJWTGuard.php %}
+```
+
+To override the existing classes when the ones you just added, you must create a [Service Provider](https://laravel.com/docs/10.x/providers) in `app/FusionAuth/Providers/FusionAuthServiceProvider.php`.
+
+```php
+{% remote_include https://raw.githubusercontent.com/FusionAuth/fusionauth-example-laravel-api/main/laravel/app/FusionAuth/Providers/FusionAuthServiceProvider.php %}
+```
+
+Finally, expose your Service Provider by adding it to the `providers` array in `config/app.php`.
+
+```diff
+{% remote_include https://github.com/FusionAuth/fusionauth-example-laravel-api/commit/b291247bc0373cffa4edc1a64a58e475f706ff40.diff %}
+```
+
+### Retrieving Public Key
+
+To make your API trust JWTs issued by FusionAuth, you must import the Public Key from your FusionAuth application into it.
```

Comment:
Hmmm. There's no support for JWKS? https://github.com/lcobucci/jwt/issues/32 or https://github.com/acodercat/php-jwk-to-pem might work. Or file an issue and offer a patch if we can extend https://github.com/tymondesigns/jwt-auth to pull the key from a known JWKS location.

I'd rather not import the JWT into

Reply:
Unfortunately no..

I spent some time trying to add JWKS support to lcobucci/jwt#32 but I couldn't make it work without changing a lot of code. Then I tried using [the lib they mentioned there](https://web-token.spomky-labs.com/the-components/key-jwk-and-key-set-jwkset/key-set-management) (which I have used in some non-Laravel projects), but I also spent so much time trying to make `tymondesigns/jwt-auth` use it that I even started creating my own stuff instead of it. It was so frustrating that eventually I just gave up and finished the article with what there is in the market :shrug:.

But I don't like importing the JWT as well and I agree that the best path here would be patching one of those libs to accept JWKS.

Reply:
Can you reach out and see if we can sponsor this? Would be happy to explore paying the contributor to support it, especially if it supports reading from a cookie as well as an authorization header.

Reply:
Otherwise, I think we should say mention as an aside that you can use a gateway as well (like kong, etc) and point there.